draft-ietf-quic-recovery.md

```diff
@@ -301,63 +302,117 @@ all subsequent ACK frames containing them could be lost. In this case, the
+not correct for any host delays reported by the peer ({{ack-delay}}).
+
+An RTT sample MUST NOT be generated for a packet that is not newly acknowledged.
+This avoids multiple RTT samples for the same packet.
+
+An RTT sample MUST NOT be generated for a packet that is not the largest
+acknowledged in the received ACK frame.  A peer reports host delays for only the
+largest acknowledged packet in an ACK frame, which is assumed by subsequent
+computations of smoothed_rtt and rttvar in correcting for host delays.
+
+An RTT sample MUST NOT be generated if the ACK frame does not newly acknowledge
+an ack-eliciting packet. A peer does not send an ACK frame in response to
+receiving only non-ack-eliciting packets, and an ACK frame that is subsequently
+sent can include an arbitrarily large Ack Delay field.  Ignoring such ACK frames
+avoids complications in subsequent smoothed_rtt and rttvar computations.
+
+A sender might generate multiple RTT samples per RTT when multiple ACK frames
+are received within an RTT.  As suggested in {{?RFC6298}}, doing so might result
+in inadequate history in smoothed_rtt and rttvar.  This is an open question
+which we leave for future research to address.
+
+## Computing min_rtt {#min-rtt}
+
+min_rtt is the minimum RTT observed over the lifetime of the connection.
+min_rtt is set to the latest_rtt on the first sample in a connection, and to the
+lesser of min_rtt and latest_rtt on subsequent samples.
+
+An endpoint uses only locally observed times in computing the min_rtt and does
+not correct for host delays reported by the peer ({{ack-delay}}).  Doing so
+allows the endpoint to set a lower bound for the smoothed_rtt based entirely on
+what it observes (see {{smoothed-rtt}}), and limits potential underestimation
+due to erroneously-reported delays by the peer.
+
+## Computing smoothed_rtt and rttvar {#smoothed-rtt}
+
+smoothed_rtt is an exponentially-weighted moving average of an endpoint's
+RTT samples, and rttvar is the endpoint's estimated variance in
+the RTT samples.
+
+smoothed_rtt estimates path latency by correcting RTT samples for peer-reported
+host delays ({{ack-delay}}).  A peer indicates that the peer's delay in sending
+an acknowledgement for an ack-eliciting packet will be no greater than the
+max_ack_delay transport parameter.  Consequently, when a peer reports an Ack
+Delay that is greater than max_ack_delay, the delay is attributed to reasons out
+of the peer's control, such as scheduler latency at the peer or loss of previous
+ACK frames.  Any delays beyond the peer's max_ack_delay are therefore considered
+effectively part of path delay and incorporated into the smoothed_rtt estimate.
+
+When correcting an RTT sample using peer-reported acknowledgement delays, an
+endpoint:
+
+- SHOULD use the lesser of the value reported in Ack Delay field of the ACK
+  frame and the peer's max_ack_delay transport parameter ({{ack-delay}}).
```

Comment:
Why SHOULD?  Why use 2119 language here at all?

> When correcting an RTT sample, endpoints first limit the reported Ack Delay to the value of the max_ack_delay transport parameter.  The RTT sample is then reduced by the resulting delay value.  If correcting for host delays would result in a sample that is less than min_rtt, the unmodified sample is used.  This guards against underestimation of RTT as a result of inflated reporting of host delays.

Reply:
Fair point.  Ultimately, this is bounded by another value that the peer advertises (max_ack_delay), so there's no additional safety issue here.